tests/test_llama.py

```diff
@@ -0,0 +1,176 @@
+                    loss = extract_loss(line)
+                    assert loss < EXPECTED_LOSS
+            # for iteration = CHECK_ITERATION_END, loss should be below EXPECTED_LOSS_END
+            if re.search(rf"iteration: {CHECK_ITERATION_END} / ", line.decode("utf-8")):
+                loss = extract_loss(line)
+                assert loss < EXPECTED_LOSS_END
+
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+
+# also run the tiny llama example. Only want to assert it can be ran.
+def test_tiny_llama():
+    # create config file
+    cmd = f"python {TINY_LLLAMA_CREATE_CONFIG_FILE}"
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    # Read and print output in real-time
+    while True:
+        line = process.stdout.readline()
+        if process.poll() is not None and line == b"":
+            break
+        if line:
+            print(line.decode("utf-8"), end="")
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+    # run training
+    # set DISABLE_FLASH_ATTENTION=1 to replace flash attention implementations
+    cmd = f'DISABLE_FLASH_ATTENTION=1 FI_PROVIDER="efa" CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun --nproc_per_node={NUM_GPUS} --rdzv_endpoint=localhost:29800  {TRAIN_SCRIPT} --config-file {TINY_LLLAMA_CONFIG_FILE}'
+    os.setpgrp()  # create new process group, become its leader
+    atexit.register(exit_with_children)  # kill all children processes when this process exits
+
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    while True:
+        line = process.stdout.readline()
+        if process.poll() is not None and line == b"":
+            break
+        if line:
+            print(line.decode("utf-8"), end="")
+
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+
+if __name__ == "__main__":
+    # create config file
+    cmd = f"python {CREATE_CONFIG_FILE}"
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    try:
+        # Read and print output in real-time
+        while True:
+            line = process.stdout.readline()
+            if process.poll() is not None and line == b"":
+                break
+            if line:
+                print(line.decode("utf-8"), end="")
+        process.wait()  # Wait for the process to finish
+        assert process.returncode == 0
+    except AssertionError:
+        print("Command failed with exit code:", process.returncode)
+        exit()
+    else:
+        print("Config created successfully.")
+
+    # run training
+    # set DISABLE_FLASH_ATTENTION=1 to replace flash attention implementations
+    cmd = f'DISABLE_FLASH_ATTENTION=1 FI_PROVIDER="efa" CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun --nproc_per_node={NUM_GPUS} --rdzv_endpoint=localhost:29800  {TRAIN_SCRIPT} --config-file {CONFIG_FILE}'
```

Comment:
Use pytest parametrize with with the parallel sizes:
- tp=1, dp=1, pp=1
- tp=2, dp=2, pp=2
- tp=1, dp=2, pp=1

Reply:
For now, the training takes 3 hours with T4. If we add more tests, I'm afraid it'll take too much time. Nouamane said that we only care about tp for now. @NouamaneTazi 
Otherwise, I can also reduce the steps to add more parameters.